chain-tx-enclave/tx-query-next/app-runner/src/zmq_helper.rs

```diff
@@ -0,0 +1,204 @@
+                inner.socket.send(message, FLAGS)?;
+                inner.state = SocketState::RequestSent;
+                Ok(Async::Ready(()))
+            }
+            _ => {
+                log::debug!("Unable to send message. Previous response isn't processed");
+                Ok(Async::NotReady)
+            }
+        }
+    }
+
+    /// Returns `true` if previous request was sent. `false` otherwise.
+    pub fn is_request_sent(&mut self) -> Result<Async<bool>> {
+        let inner = lock!(self.inner);
+        Ok(Async::Ready(matches!(
+            inner.state,
+            SocketState::RequestSent
+        )))
+    }
+
+    /// Returns `true` if response was received. `false` otherwise.
+    pub fn is_response_received(&mut self) -> Result<Async<bool>> {
+        let inner = lock!(self.inner);
+        Ok(Async::Ready(matches!(
+            inner.state,
+            SocketState::ResponseReceived(_)
+        )))
+    }
+
+    /// Returns the length of message received from ZeroMQ
+    pub fn get_message_len(&mut self) -> Result<Async<usize>> {
+        let mut inner = lock!(self.inner);
+
+        let message = match inner.state {
+            SocketState::RequestSent => inner.socket.recv_msg(FLAGS)?,
+            _ => {
+                log::debug!("Unable to receive message. Previous request wasn't sent");
+                return Ok(Async::NotReady);
+            }
+        };
+
+        let message_len = message.len();
+        inner.state = SocketState::ResponseReceived(message);
+
+        Ok(Async::Ready(message_len))
+    }
+
+    /// Returns the message received from ZeroMQ
+    pub fn get_message(&mut self) -> Result<Async<Message>> {
+        log::debug!("Receiving response to ZeroMQ");
+        let mut inner = lock!(self.inner);
+
+        if matches!(inner.state, SocketState::ResponseReceived(_)) {
+            let mut socket_state = SocketState::Ready;
+            std::mem::swap(&mut socket_state, &mut inner.state);
+
+            if let SocketState::ResponseReceived(message) = socket_state {
+                Ok(Async::Ready(message))
+            } else {
+                unreachable!("Socket state cannot be anything other than `ResponseReceived`")
+            }
+        } else {
+            log::debug!("Unable to receive message. Previous request wasn't sent");
+            Ok(Async::NotReady)
+        }
+    }
+}
+
+impl Read for ZmqHelper {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if try_poll!(self.is_request_sent()?) {
+            let message_len: u32 = try_poll!(self.get_message_len()?)
+                .try_into()
+                .expect("Message length exceeds `u32` bounds");
+
+            buf.copy_from_slice(&message_len.to_le_bytes());
+            Ok(core::mem::size_of::<u32>())
+        } else if try_poll!(self.is_response_received()?) {
+            let message = try_poll!(self.get_message()?);
+            buf.copy_from_slice(&message);
```

Comment:
It'll be good to document this extra semantics on top of / from ZmqHelper, as one may get panics from `copy_from_slice` (e.g. if one uses `[u8; 8]` for the first buffer?)